File: Algorithms/p162_Find_Peak_Element/p162_Find_Peak_Element_2.py
class Solution(object):
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        l = 0
        r = len(nums)
        while True:
            mid = (l+r)//2
            if (mid==0 or nums[mid]>nums[mid-1]) and (mid==len(nums)-1 or nums[mid]>nums[mid+1]):
                return mid
            elif (mid==0 or nums[mid]>nums[mid-1]) and (mid<len(nums)-1 and nums[mid]<nums[mid+1]):
                l = mid+1
            else:
                r = mid
        return -1 # dummy return

File: Algorithms/p162_Find_Peak_Element/p162_Find_Peak_Element_2.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int l = 0;
        int r = nums.size();
        while(1){
            int mid = (l+r)/2;
            if((mid==0 || nums[mid]>nums[mid-1]) && (mid==nums.size()-1 || nums[mid]>nums[mid+1])){
                return mid;
            }else if((mid==0 || nums[mid]>nums[mid-1]) && (mid<nums.size()-1 && nums[mid]<nums[mid+1])){
                l = mid+1;
            }else{
                r = mid;
            }
        }
        return -1; // dummy return
    }
};
